src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -165,19 +162,46 @@ internal void AppendMetadata()
                 }
 
                 WriteCustomAttributes(typeBuilder._customAttributes, typeHandle);
-                WriteMethods(typeBuilder);
+                WriteMethods(typeBuilder, genericParams);
                 WriteFields(typeBuilder);
             }
+
+            // Now write all generic parameters in order
+            genericParams.Sort(GenericParameterComparer);
```

Comment:
This will allocate a new delegate every time this is called. The lambda from my suggestion would be a cached delegate.